File: 0409-longest-palindrome/0409-longest-palindrome.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        freq = {}

        # Count frequency
        for ch in s:
            freq[ch] = freq.get(ch, 0) + 1

        length = 0
        hasOdd = False

        for count in freq.values():
            length += (count // 2) * 2

            if count % 2 == 1:
                hasOdd = True

        if hasOdd:
            length += 1

        return length

File: 0409-longest-palindrome/0409-longest-palindrome.java
class Solution {
    public int longestPalindrome(String s) {

        HashMap<Character, Integer> map = new HashMap<>();

        // Count frequency of each character
        for (char ch : s.toCharArray()) {
            map.put(ch, map.getOrDefault(ch, 0) + 1);
        }

        int length = 0;
        boolean hasOdd = false;

        // Calculate longest palindrome length
        for (int count : map.values()) {

            if (count % 2 == 0) {
                length += count;
            } else {
                length += count - 1; // Use even part
                hasOdd = true;
            }
        }

        // Add one odd character in the center
        if (hasOdd) {
            length++;
        }

        return length;
    }
}